### PR TITLE
Deduplication PR

### PR DESCRIPTION
**Pull request by hexylena**
Original PR message:

> I ran a code duplication tool over `lib/` and found some pretty awful copy-pasta. This PR cleans that up a little. It's probably a sign of a bigger architectural issue and this is just a bandaid.
> 
> NB: I have not manually tested the code to verify it still functions correctly, so feel free to decline this PR on those grounds, if no one else is interested in testing.

I'm expanding this to include other deduplication changes. They should probably get some more testing. If you're reviewing this, I strongly recommend viewing it commit-by-commit to understand where things are getting refactored from/to

### CODE REVIEW

File: lib/galaxy/webapps/tool_shed/controllers/repository.py [hunk 1]
```diff
@@ -157,21 +174,10 @@ def browse_deprecated_repositories_i_own( self, trans, **kwd ):
...
-            elif operation == "repositories_by_user":
-                return trans.response.send_redirect( web.url_for( controller='repository',
-                                                                  action='browse_repositories_by_user',
-                                                                  **kwd ) )
-            elif operation in [ 'mark as deprecated', 'mark as not deprecated' ]:
-                kwd[ 'mark_deprecated' ] = operation == 'mark as deprecated'
-                return trans.response.send_redirect( web.url_for( controller='repository',
-                                                                  action='deprecate',
-                                                                  **kwd ) )
+        _redir = self._redirect_if_necesseary( trans, **kwd )
```
Comment (martenson) [inline lib/galaxy/webapps/tool_shed/controllers/repository.py]: copy pasta typo (:

Comment (hexylena) [inline lib/galaxy/webapps/tool_shed/controllers/repository.py]: heh, indeed. Nice catch.

**Comment by hexylena** [comment]
Okay, that's the bulk of the _major_ duplications (other than repository_grids.py, but that's not worth the time investment). 
- pulsar/lwr are highly duplicated
- Other things require extensive work to refactor/are architectural issues

Removing WIP tag as that's all the time I have for this right now.

File: lib/galaxy/security/__init__.py [hunk 2]
```diff
@@ -208,11 +214,7 @@ def get_all_roles( self, trans, cntrller ):
             for role in self.get_sharing_roles( trans.user ):
                 roles.add( role )
             # Add all remaining non-private, non-sharing roles
-            for role in trans.sa_session.query( trans.app.model.Role ) \
-                                        .filter( and_( self.model.Role.table.c.deleted == false(),
-                                                       self.model.Role.table.c.type != self.model.Role.types.PRIVATE,
-                                                       self.model.Role.table.c.type != self.model.Role.types.SHARING ) ) \
-                                        .order_by( self.model.Role.table.c.name ):
+            for role in self._get_roles( trans):
```
Comment (nsoranzo) [inline lib/galaxy/security/__init__.py]: s/_get_roles/_get_npns_roles/

File: lib/galaxy/webapps/util.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+def build_template_error_formatters():
+    """
+    Build a list of template error formatters for WebError. When an error
+    occurs, WebError pass the exception to each function in this list until
+    one returns a value, which will be displayed on the error page.
+    """
+    formatters = []
+    # Formatter for mako
+    import mako.exceptions
```
Comment (nsoranzo) [inline lib/galaxy/webapps/util.py]: Can you please move this at the top of the file, preceded by:

``` python
from galaxy import eggs
eggs.require("Mako")
```

File: lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py [hunk 4]
```diff
@@ -65,27 +67,10 @@ def create_api_key( app, user ):
...
     app.sa_session.flush()
     return api_key
-    
+
+
 def create_user( app ):
-    user_info_config = os.path.abspath( os.path.join( app.config.root, 'lib/tool_shed/scripts/bootstrap_tool_shed', 'user_info.xml' ) )
-    email = None
-    password = None
-    username = None
```
Comment (nsoranzo) [inline lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py:68]: I think these 3 initialization lines should be added to `.bootstrap_util.admin_user_info()`.

Comment (hexylena) [inline lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py:68]: Okay. Can do. There were two equivalent implementations so I picked the cleanest one. I can see a case where it wouldn't set those variables, so I'll copy the lines.

**Comment by nsoranzo** [comment]
:+1: 

Thanks @erasche!

"194 additions and 450 deletions" = Cool!

**Comment by jmchilton** [comment]
Pushed a conflict resolution merge for this. :+1: on this, will merge later today unless someone complains. Also thanks Eric!!

**Comment by jmchilton** [comment]
About the Pulsar/LWR duplication - that is largely intentional. The LWR code in Galaxy will sit unmodified until finally removed. Modifying any of it does not make sense to me, it might break something and it is all deprecated. It might be time to remove the LWR code altogether just to clean up stuff though, I'll post to -dev about it.

**Comment by hexylena** [comment]
@jmchilton thanks for the conflict-resolution merge! 

Yep, figured that it was intentional and thanks for posting to dev. Would be nice to get rid of legacy cruft :)

**Pull request merged**